bam_consensus.c

```diff
@@ -380,6 +384,49 @@ int nins(const bam1_t *b){
     return indel;
 }
 
+/*
+ * Some machines, including 454 and PacBio, store the quality values in
+ * homopolymers with the first or last base always being the low quality
+ * state.  This can cause problems when reverse-complementing and aligning,
+ * especially when we left-justify indels.
+ *
+ * Other platforms take the approach of having the middle bases high and
+ * the low confidence spread evenly to both start and end.  This means
+ * reverse-complementing doesn't introduce any strand bias.
+ *
+ * We redistribute qualities within homopolymers in this style to fix
+ * naive consensus or variant calling algorithms.
+ */
+void homopoly_qual_fix(bam1_t *b) {
+    static double ph2err[256] = {0};
+    if (!ph2err[0]) {
+        for (int i = 0; i < 256; i++)
+            ph2err[i] = pow(10, i/-10.0);
+    }
+    uint8_t *seq = bam_get_seq(b);
+    uint8_t *qual = bam_get_qual(b);
+    for (int i = 0; i < b->core.l_qseq; i++) {
+        int s = i; // start of homopoly
+        int base = bam_seqi(seq, i);
+        while (i+1 < b->core.l_qseq && bam_seqi(seq, i+1) == base)
+            i++;
+        // s..i inclusive is now homopolymer
+
+        if (s == i)
+            continue;
+
+        // Simplest:  reverse if end_qual < start_qual
+        // Next:      average outer-most two, then next two, etc
+        // Best:      fully redistribute so start/end lower qual than centre
+
+        // Middle route of averaging outer pairs is sufficient?
+        for (int j = s, k = i; j < k; j++,k--) {
+            double e = ph2err[qual[j]] + ph2err[qual[k]];
+            qual[j] = qual[k] = -fast_log2(e/2)*3.0104+.49;
+        }
```

Comment:
The result of this transformation is that the new quality value is the minimum of the original two, plus a factor which varies between 0 and 3 depending on the absolute difference of the qualities.  So, if I've got this right the same result could be had without the floating-point manipulations like this:

```
uint8_t adjust[] = { 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 3, 3, /* 3's omitted ... */, 3 };
qual[j] = qual[k] = MIN(qual[j], qual[k]) + adjust(ABS(qual[j] - qual[k]));
```

Adding three is essentially halving the probability that the base call is incorrect (and arises from the `/2` in the original formulation).

Presumably if there's a big drop along the length of the run, the quality at the high end goes down by quite a lot.  Could this be the reason for the slight increase in false negatives you saw in your tests?

Reply:
Yes, it's basically +3 in the limit of large difference due to it being phred for 0.5.  I did think about doing that, but the above code was clearer and `fast_log2` wasn't a bottleneck that I could see.

Note this isn't a drop along the length of a run (that was one idea I had and is in the comments above), but rather swapping ends only.  However potentially there is benefit to a more complex analysis, e.g. when TAAAAAC is mutated to TAAGAAC.  It's low hanging fruit stuff though.  An alternative is sorting instead of averaging, e.g. left normalisatoin.  I haven't explored that though and my gut feeling is it may make problems elsewhere.

I did also wonder about a slight rescaling of the resulting qual values (for all cases obviously, not just homopolymers) so we have similar FN rates and perhaps rescue the lowest qual ones.  However none of this is particularly well calibrated in the first place and QUAL filtering is therefore somewhat arbitrary.  It may still be something to consider though.

Reply:
I should also add in the spirit of more smaller PRs instead of large ones, this is part 1.

Part 2 revises the consensus algorithm itself, instead of the quality values supplied to it.  It's based on the later platform-specific indel rates added to the gap5 consensus algorithm.  This has a significant impact on calling, but it still greatly benefits from the simple -p transformation here.